pooltool/physics/resolve/stronge_compliant.py

```diff
@@ -0,0 +1,516 @@
+    return -mu * s * f_per_m_n(time, compression_duration, v_n_0, beta_n, omega_n, e_n)
+
+
+@jit(nopython=True, cache=const.use_numba_cache)
+def f_per_m_t_intermediate_angle_of_incidence(
+    time,
+    compression_duration,
+    v_n_0,
+    beta_t,
+    beta_n,
+    omega_t,
+    omega_n,
+    mu,
+    e_n,
+    v_t_stick,
+    u_t_stick,
+    time_stick,
+    time_slip,
+):
+    if time <= time_stick:
+        s = -1
+    elif time >= time_slip:
+        s = 1
+    else:
+        s = 0
+    return (
+        f_per_m_t_slip(time, compression_duration, v_n_0, beta_n, omega_n, mu, e_n, s)
+        if s != 0
+        else f_per_m_t_stick(time, beta_t, omega_t, v_t_stick, u_t_stick, time_stick)
+    )
+
+
+@jit(nopython=True, cache=const.use_numba_cache)
+def is_initial_stick(v_t_0_by_v_n_0, mu, eta_squared):
+    return v_t_0_by_v_n_0 < mu * eta_squared
+
+
+@jit(nopython=True, cache=const.use_numba_cache)
+def is_gross_slip(v_t_0_by_v_n_0, mu, e_n, eta_squared, beta_t_by_beta_n):
+    return v_t_0_by_v_n_0 > mu * ((1 + e_n) * beta_t_by_beta_n - eta_squared / e_n)
+
+
+def slip_time_for_initial_stick(
+    v_t_0_by_v_n_0: float,
+    omega_t: float,
+    omega_n: float,
+    mu: float,
+    e_n: float,
+    eta_squared: float,
+    t_c: float,
+    t_f: float,
+) -> float:
+    def f(t):
+        return abs(-v_t_0_by_v_n_0 * math.sin(omega_t * t)) - (
+            mu
+            * eta_squared
+            * omega_t
+            / omega_n
+            * math.sin(phase_angle_restitution(t, omega_n, e_n))
+        )
+
+    if np.isclose(v_t_0_by_v_n_0, 0):
+        return t_f
+    if np.isclose(v_t_0_by_v_n_0, mu * eta_squared):
+        return t_c
+
+    t_slip = sp.optimize.toms748(f, t_c, t_f)
+    assert t_c <= t_slip and t_slip <= t_f, f"t_c={t_c} <= t_slip={t_slip} <= t_f={t_f}"
+    return t_slip
+
+
+def slip_time_for_stick(
+    v_t_0: float,
+    v_n_0: float,
+    beta_t_by_beta_n: float,
+    omega_t: float,
+    omega_n: float,
+    mu: float,
+    e_n: float,
+    eta_squared: float,
+    t_c: float,
+    t_f: float,
+    t_stick: float = 0,
+    u_t_stick: float = 0,
+    v_t_stick: float = 0,
+) -> float:
+    v_t_0_by_v_n_0 = v_t_0 / v_n_0
+
+    def f(t):
+        return abs(
+            omega_n
+            / (mu * v_n_0)
+            * (
+                u_t_stick * math.cos(omega_t * (t - t_stick))
+                - v_t_stick / omega_t * math.sin(omega_t * (t - t_stick))
+            )
+        ) - eta_squared * math.sin(phase_angle_restitution(t, omega_n, e_n))
+
+    if np.isclose(v_t_0_by_v_n_0, mu * eta_squared):
+        return t_c
+    if np.isclose(
+        v_t_0_by_v_n_0, mu * ((1 + e_n) * beta_t_by_beta_n - eta_squared / e_n)
+    ):
+        return t_f
+
+    t_slip = sp.optimize.toms748(f, t_c, t_f)
+    """
+    assert t_stick <= t_slip and t_slip <= t_f, (
+        f"t_stick={t_stick} <= t_slip={t_slip} <= t_f={t_f}"
+    )
+    """
```

Comment:
This is evaluated as a string at runtime. Probably best to remove or keep the assert.